File: ios/singword/UI/Search/SearchViewModel.swift
import Combine
import Foundation

enum SearchErrorCode {
    case none
    case emptyQuery
    case noWordbookSelected
    case wordbookMissingAsset
    case wordbookParseError
    case lyricsNotFound
    case networkError
    case providerError
    case unknown
}

struct SearchUiState {
    var query: String = ""
    var isLoading: Bool = false
    var trackName: String = ""
    var artistName: String = ""
    var provider: String = ""
    var candidates: [LyricsCandidate] = []
    var selectedCandidate: LyricsCandidate?
    var matchedWords: [MatchedWord] = []
    var totalTokens: Int = 0
    var isEmptyResult: Bool = false
    var error: String?
    var errorCode: SearchErrorCode = .none
}

@MainActor
final class SearchViewModel: ObservableObject {
    @Published private(set) var uiState = SearchUiState()
    @Published private(set) var favoriteWords: Set<String> = []

    private let lyricsRepository: LyricsRepository
    private let wordbookRepository: WordbookRepository
    private let settingsRepository: SettingsRepository
    private let favoritesStore: FavoritesStore
    private var cancellables: Set<AnyCancellable> = []
    private var activeTask: Task<Void, Never>?

    init(
        lyricsRepository: LyricsRepository,
        wordbookRepository: WordbookRepository,
        settingsRepository: SettingsRepository,
        favoritesStore: FavoritesStore
    ) {
        self.lyricsRepository = lyricsRepository
        self.wordbookRepository = wordbookRepository
        self.settingsRepository = settingsRepository
        self.favoritesStore = favoritesStore

        favoritesStore.$favoriteWords
            .sink { [weak self] words in
                self?.favoriteWords = words
            }
            .store(in: &cancellables)
    }

    func onQueryChanged(_ query: String) {
        uiState.query = query
    }

    func searchCandidates() -> Bool {
        let query = uiState.query.trimmingCharacters(in: .whitespacesAndNewlines)
        if query.isEmpty {
            setError(message: "请输入歌名", code: .emptyQuery)
            return false
        }

        if settingsRepository.getEnabledWordbooks().isEmpty {
            setError(message: "请先在设置中选择至少一个词表", code: .noWordbookSelected)
            return false
        }

        resetForNewSearch()

        runOperation { [self] in
            switch await lyricsRepository.searchCandidates(query, limit: 5) {
            case .success(let candidates, let provider):
                guard !Task.isCancelled else { return }
                uiState.isLoading = false
                uiState.provider = provider
                uiState.candidates = candidates
                uiState.error = nil
                uiState.errorCode = .none

            case .notFound:
                guard !Task.isCancelled else { return }
                setError(message: "未找到歌词", code: .lyricsNotFound)

            case .networkError(let message):
                guard !Task.isCancelled else { return }
                setError(message: message, code: .networkError)

            case .providerError(_, let message):
                guard !Task.isCancelled else { return }
                setError(message: message, code: .providerError)
            }
        }

        return true
    }

    func selectCandidate(_ candidate: LyricsCandidate) {
        runOperation { [self] in
            await matchCandidate(candidate)
        }
    }

    func search() {
        let query = uiState.query.trimmingCharacters(in: .whitespacesAndNewlines)
        if query.isEmpty {
            setError(message: "请输入歌名", code: .emptyQuery)
            return
        }

        resetForNewSearch()

        runOperation { [self] in
            switch await lyricsRepository.searchLyrics(query) {
            case .success(let trackName, let artistName, let lyrics, let provider):
                guard !Task.isCancelled else { return }
                await matchCandidate(
                    LyricsCandidate(
                        trackName: trackName,
                        artistName: artistName,
                        lyrics: lyrics,
                        provider: provider
                    )
                )

            case .notFound:
                guard !Task.isCancelled else { return }
                setError(message: "未找到歌词", code: .lyricsNotFound)

            case .networkError(let message):
                guard !Task.isCancelled else { return }
                setError(message: message, code: .networkError)

            case .providerError(_, let message):
                guard !Task.isCancelled else { return }
                setError(message: message, code: .providerError)
            }
        }
    }

    func retryResult() {
        if let selected = uiState.selectedCandidate {
            selectCandidate(selected)
        } else {
            search()
        }
    }

    func clearError() {
        uiState.error = nil
        uiState.errorCode = .none
    }

    func toggleFavorite(_ word: MatchedWord) {
        Task {
            await favoritesStore.toggle(word)
        }
    }

    private func loadEnabledWordbooksOrError() async -> [String: (WordEntry, String)]? {
        let enabled = settingsRepository.getEnabledWordbooks()
        if enabled.isEmpty {
            setError(message: "请先在设置中选择至少一个词表", code: .noWordbookSelected)
            return nil
        }

        switch await wordbookRepository.loadEnabledWordbooks(enabled: enabled) {
        case .success(let words):
            return words

        case .missingAsset(let assetPath):
            setError(message: "词表文件缺失：\(assetPath)", code: .wordbookMissingAsset)
            return nil

        case .parseError(let assetPath, _):
            setError(message: "词表解析失败：\(assetPath)", code: .wordbookParseError)
            return nil
        }
    }

    private func matchCandidate(_ candidate: LyricsCandidate) async {
        guard !Task.isCancelled else { return }
        uiState.isLoading = true
        uiState.trackName = candidate.trackName
        uiState.artistName = candidate.artistName
        uiState.provider = candidate.provider
        uiState.selectedCandidate = candidate
        uiState.matchedWords = []
        uiState.totalTokens = 0
        uiState.isEmptyResult = false
        uiState.error = nil
        uiState.errorCode = .none

        guard let wordbooks = await loadEnabledWordbooksOrError() else {
            guard !Task.isCancelled else { return }
            uiState.isLoading = false
            return
        }

        guard !Task.isCancelled else { return }
        let tokens = LyricsProcessor.tokenize(candidate.lyrics)
        let matchedWords = VocabMatcher.match(tokens: tokens, wordbooks: wordbooks)

        guard !Task.isCancelled else { return }
        uiState.isLoading = false
        uiState.trackName = candidate.trackName
        uiState.artistName = candidate.artistName
        uiState.provider = candidate.provider
        uiState.selectedCandidate = candidate
        uiState.matchedWords = matchedWords
        uiState.totalTokens = tokens.count
        uiState.isEmptyResult = matchedWords.isEmpty
        uiState.error = nil
        uiState.errorCode = .none
    }

    private func resetForNewSearch() {
        uiState.isLoading = true
        uiState.trackName = ""
        uiState.artistName = ""
        uiState.provider = ""
        uiState.candidates = []
        uiState.selectedCandidate = nil
        uiState.matchedWords = []
        uiState.totalTokens = 0
        uiState.isEmptyResult = false
        uiState.error = nil
        uiState.errorCode = .none
    }

    private func setError(message: String, code: SearchErrorCode) {
        uiState.isLoading = false
        uiState.error = message
        uiState.errorCode = code
    }

    private func runOperation(_ operation: @escaping () async -> Void) {
        activeTask?.cancel()
        activeTask = Task { [weak self] in
            guard self != nil else { return }
            await operation()
        }
    }
}

File: android/app/src/main/java/com/singword/app/ui/search/SearchViewModel.kt
package com.singword.app.ui.search

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.singword.app.data.local.db.FavoriteRepository
import com.singword.app.data.local.db.FavoriteWord
import com.singword.app.data.local.prefs.SettingsRepository
import com.singword.app.data.local.song.DownloadedSongRepository
import com.singword.app.data.local.song.RecentSearchRepository
import com.singword.app.data.local.song.SongMatchSnapshot
import com.singword.app.data.local.widget.WidgetSnapshotStore
import com.singword.app.data.local.wordbook.WordEntry
import com.singword.app.data.local.wordbook.WordbookLoadResult
import com.singword.app.data.local.wordbook.WordbookRepository
import com.singword.app.data.remote.LyricsCandidate
import com.singword.app.data.remote.LyricsCandidateResult
import com.singword.app.data.remote.LyricsRepository
import com.singword.app.data.remote.LyricsResult
import com.singword.app.domain.LyricsProcessor
import com.singword.app.domain.MatchedWord
import com.singword.app.domain.VocabMatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

enum class SearchErrorCode {
    NONE,
    EMPTY_QUERY,
    NO_WORDBOOK_SELECTED,
    WORDBOOK_MISSING_ASSET,
    WORDBOOK_PARSE_ERROR,
    LYRICS_NOT_FOUND,
    NETWORK_ERROR,
    PROVIDER_ERROR,
    UNKNOWN
}

data class SearchUiState(
    val query: String = "",
    val isLoading: Boolean = false,
    val trackName: String = "",
    val artistName: String = "",
    val provider: String = "",
    val candidates: List<LyricsCandidate> = emptyList(),
    val selectedCandidate: LyricsCandidate? = null,
    val matchedWords: List<MatchedWord> = emptyList(),
    val totalTokens: Int = 0,
    val isEmptyResult: Boolean = false,
    val error: String? = null,
    val errorCode: SearchErrorCode = SearchErrorCode.NONE
)

class SearchViewModel(
    private val lyricsRepository: LyricsRepository,
    private val wordbookRepository: WordbookRepository,
    private val settingsRepository: SettingsRepository,
    private val favoriteRepository: FavoriteRepository,
    private val recentSearchRepository: RecentSearchRepository,
    private val downloadedSongRepository: DownloadedSongRepository,
    private val widgetSnapshotStore: WidgetSnapshotStore
) : ViewModel() {

    private val _uiState = MutableStateFlow(SearchUiState())
    val uiState: StateFlow<SearchUiState> = _uiState.asStateFlow()

    val favoriteWords: StateFlow<Set<String>> = favoriteRepository.getAllFavoriteWords()
        .map { it.toSet() }
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptySet())

    val recentSearches: StateFlow<List<SongMatchSnapshot>> = recentSearchRepository.getAll()
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    val downloadedSongIds: StateFlow<Set<String>> = downloadedSongRepository.getAllIds()
        .map { it.toSet() }
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptySet())

    fun onQueryChanged(query: String) {
        _uiState.update { it.copy(query = query) }
    }

    fun searchCandidates(): Boolean {
        val query = _uiState.value.query.trim()
        if (query.isBlank()) {
            setError(
                message = "请输入歌名",
                code = SearchErrorCode.EMPTY_QUERY
            )
            return false
        }

        if (settingsRepository.getEnabledWordbooks().isEmpty()) {
            setError(
                message = "请先在设置中选择至少一个词表",
                code = SearchErrorCode.NO_WORDBOOK_SELECTED
            )
            return false
        }

        viewModelScope.launch(Dispatchers.IO) {
            _uiState.update {
                it.copy(
                    isLoading = true,
                    trackName = "",
                    artistName = "",
                    provider = "",
                    candidates = emptyList(),
                    selectedCandidate = null,
                    matchedWords = emptyList(),
                    totalTokens = 0,
                    isEmptyResult = false,
                    error = null,
                    errorCode = SearchErrorCode.NONE
                )
            }

            when (val candidateResult = lyricsRepository.searchCandidates(query, limit = 5)) {
                is LyricsCandidateResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            provider = candidateResult.provider,
                            candidates = candidateResult.candidates,
                            error = null,
                            errorCode = SearchErrorCode.NONE
                        )
                    }
                }
                is LyricsCandidateResult.NotFound -> {
                    setError(
                        message = "未找到歌词",
                        code = SearchErrorCode.LYRICS_NOT_FOUND
                    )
                }
                is LyricsCandidateResult.NetworkError -> {
                    setError(
                        message = candidateResult.message,
                        code = SearchErrorCode.NETWORK_ERROR
                    )
                }
                is LyricsCandidateResult.ProviderError -> {
                    setError(
                        message = candidateResult.message,
                        code = SearchErrorCode.PROVIDER_ERROR
                    )
                }
            }
        }

        return true
    }

    fun selectCandidate(candidate: LyricsCandidate) {
        viewModelScope.launch(Dispatchers.IO) {
            matchCandidate(candidate)
        }
    }

    fun search() {
        val query = _uiState.value.query.trim()
        if (query.isBlank()) {
            setError(
                message = "请输入歌名",
                code = SearchErrorCode.EMPTY_QUERY
            )
            return
        }

        viewModelScope.launch(Dispatchers.IO) {
            _uiState.update {
                it.copy(
                    isLoading = true,
                    trackName = "",
                    artistName = "",
                    provider = "",
                    candidates = emptyList(),
                    selectedCandidate = null,
                    matchedWords = emptyList(),
                    totalTokens = 0,
                    isEmptyResult = false,
                    error = null,
                    errorCode = SearchErrorCode.NONE
                )
            }

            when (val lyricsResult = lyricsRepository.searchLyrics(query)) {
                is LyricsResult.Success -> {
                    matchCandidate(
                        LyricsCandidate(
                            trackName = lyricsResult.trackName,
                            artistName = lyricsResult.artistName,
                            lyrics = lyricsResult.lyrics,
                            provider = lyricsResult.provider
                        )
                    )
                }
                is LyricsResult.NotFound -> {
                    setError(
                        message = "未找到歌词",
                        code = SearchErrorCode.LYRICS_NOT_FOUND
                    )
                }
                is LyricsResult.NetworkError -> {
                    setError(
                        message = lyricsResult.message,
                        code = SearchErrorCode.NETWORK_ERROR
                    )
                }
                is LyricsResult.ProviderError -> {
                    setError(
                        message = lyricsResult.message,
                        code = SearchErrorCode.PROVIDER_ERROR
                    )
                }
            }
        }
    }

    fun retryResult() {
        val candidate = _uiState.value.selectedCandidate
        if (candidate != null) {
            selectCandidate(candidate)
        } else {
            search()
        }
    }

    fun clearError() {
        _uiState.update {
            it.copy(error = null, errorCode = SearchErrorCode.NONE)
        }
    }

    fun toggleFavorite(word: MatchedWord) {
        viewModelScope.launch(Dispatchers.IO) {
            val current = favoriteWords.value
            if (word.word in current) {
                favoriteRepository.delete(
                    FavoriteWord(
                        word = word.word,
                        pos = word.pos,
                        def = word.def,
                        source = word.source
                    )
                )
            } else {
                favoriteRepository.upsert(
                    FavoriteWord(
                        word = word.word,
                        pos = word.pos,
                        def = word.def,
                        source = word.source
                    )
                )
            }
        }
    }

    fun loadSnapshot(snapshot: SongMatchSnapshot) {
        widgetSnapshotStore.save(snapshot)
        _uiState.update {
            it.copy(
                query = snapshot.trackName,
                isLoading = false,
                trackName = snapshot.trackName,
                artistName = snapshot.artistName,
                provider = snapshot.provider,
                candidates = emptyList(),
                selectedCandidate = null,
                matchedWords = snapshot.toMatchedWords(),
                totalTokens = snapshot.totalTokens,
                isEmptyResult = snapshot.matchedWords.isEmpty(),
                error = null,
                errorCode = SearchErrorCode.NONE
            )
        }
    }

    fun downloadCurrentSong() {
        val snapshot = currentSnapshot() ?: return
        viewModelScope.launch(Dispatchers.IO) {
            downloadedSongRepository.upsert(snapshot)
        }
    }

    fun isCurrentSongDownloaded(): Boolean {
        val snapshot = currentSnapshot() ?: return false
        return snapshot.id in downloadedSongIds.value
    }

    private fun loadEnabledWordbooksOrError(): Map<String, Pair<WordEntry, String>>? {
        val enabled = settingsRepository.getEnabledWordbooks()
        if (enabled.isEmpty()) {
            setError(
                message = "请先在设置中选择至少一个词表",
                code = SearchErrorCode.NO_WORDBOOK_SELECTED
            )
            return null
        }

        return when (val loadResult = wordbookRepository.loadEnabledWordbooks(enabled)) {
            is WordbookLoadResult.Success -> loadResult.words
            is WordbookLoadResult.MissingAsset -> {
                setError(
                    message = "词表文件缺失：${loadResult.assetPath}",
                    code = SearchErrorCode.WORDBOOK_MISSING_ASSET
                )
                null
            }
            is WordbookLoadResult.ParseError -> {
                setError(
                    message = "词表解析失败：${loadResult.assetPath}",
                    code = SearchErrorCode.WORDBOOK_PARSE_ERROR
                )
                null
            }
        }
    }

    private fun matchCandidate(candidate: LyricsCandidate) {
        _uiState.update {
            it.copy(
                isLoading = true,
                trackName = candidate.trackName,
                artistName = candidate.artistName,
                provider = candidate.provider,
                selectedCandidate = candidate,
                matchedWords = emptyList(),
                totalTokens = 0,
                isEmptyResult = false,
                error = null,
                errorCode = SearchErrorCode.NONE
            )
        }

        val wordbooks = loadEnabledWordbooksOrError() ?: return

        val tokens = LyricsProcessor.tokenize(candidate.lyrics)
        val matched = VocabMatcher.match(tokens, wordbooks)
        val snapshot = SongMatchSnapshot.fromMatchedWords(
            trackName = candidate.trackName,
            artistName = candidate.artistName,
            provider = candidate.provider,
            totalTokens = tokens.size,
            matchedWords = matched
        )
        widgetSnapshotStore.save(snapshot)
        viewModelScope.launch(Dispatchers.IO) {
            recentSearchRepository.upsert(snapshot)
        }
        _uiState.update {
            it.copy(
                isLoading = false,
                trackName = candidate.trackName,
                artistName = candidate.artistName,
                provider = candidate.provider,
                selectedCandidate = candidate,
                matchedWords = matched,
                totalTokens = tokens.size,
                isEmptyResult = matched.isEmpty(),
                error = null,
                errorCode = SearchErrorCode.NONE
            )
        }
    }

    private fun setError(message: String, code: SearchErrorCode) {
        _uiState.update {
            it.copy(
                isLoading = false,
                error = message,
                errorCode = code
            )
        }
    }

    private fun currentSnapshot(): SongMatchSnapshot? {
        val state = _uiState.value
        if (state.trackName.isBlank() || state.matchedWords.isEmpty()) return null
        return SongMatchSnapshot.fromMatchedWords(
            trackName = state.trackName,
            artistName = state.artistName,
            provider = state.provider,
            totalTokens = state.totalTokens,
            matchedWords = state.matchedWords
        )
    }
}
